packages/ai/src/workflows/Workflow.ts

```diff
@@ -133,6 +140,24 @@ export class Workflow {
     return step;
   }
 
+  addRetryOfPreviousStep() {
+    const lastStep = this.steps.at(-1);
+    if (!lastStep) return;
+
+    const retryingStep = lastStep.retryingStep || lastStep;
+    const retryAttempts = this.getCurrentRetryAttempts(retryingStep.id);
+
+    if (retryAttempts >= MAX_RETRIES) {
+      return;
+    }
+
+    this.addStep(retryingStep.template, retryingStep.inputs, retryingStep);
+  }
+
+  public getCurrentRetryAttempts(stepId: string): number {
+    return this.steps.filter((step) => step.retryingStep?.id === stepId).length;
```

Comment:
Another way to count retries would be to put the retries in a chain, point `retryingStep` to the previous retry instead of the first one, and count the chain length recursively.

That might be a bit cleaner because then you don't have to refer to `workflow.steps` at all, and `getRetryAttempts` could be self-contained in `LLMStep` code.

The difference is pretty subtle and probably not very important.

Reply:
Yea, I'm not sure here. Kind of want to keep it now, then change if it seems good to do so later.